### ISSUE TITLE
Modify our code to initialize woodwork with partial schemas

### ISSUE THREAD

**Issue by freddyaboulton** [alteryx/evalml#2744]
Woodwork introduced the ability to initialize typing information with a partial schema in https://github.com/alteryx/woodwork/pull/1100

This means that we can potentially get rid of `_retain_custom_types_and_intialize_woodwork` and therefore retain more than just the logical types after some components modify the input data in a way that breaks the woodwork schema.

For example, this is how the `TargetEncoder` would look:

```python
    X_t_df = pd.DataFrame(X_t, columns=X_ww.columns, index=X_ww.index)
    cat_columns = X_ww.ww.select("categorical", return_schema=True).columns
    columns = [c for c in X_t_df.columns if c not in cat_columns]
    # Infer types on newly created columns
    X_t_df.ww.init(schema=original_schema._get_subset_schema(columns))
```

**Comment by chukarsten** [comment]
As acceptance criteria for completing this issue, let's evaluate the ability to remove `_retain_custom_types_and_intialize_woodwork` and refactor using the new WW capabilities.

**Comment by chukarsten** [comment]
@freddyaboulton do we need to look at the fix in [#2752](https://app.zenhub.com/workspaces/evalml-5d7698439dde650001eaf87c/issues/alteryx/evalml/2752) to refactor to use this as well?

**Comment by freddyaboulton** [comment]
@chukarsten If you're asking if we're going to have to revisit the `Imputer` as part of this refactor, then you're spot on. I think anywhere `_retain_custom_types_and_initalize_woodwork` is in scope for this issue!
